wikitextprocessor/parser.py

```diff
@@ -334,11 +334,40 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return self.__str__()
 
+    def find_child(self, target_kind: NodeKind) -> Iterator["WikiNode"]:
+        for child in self.children:
+            if isinstance(child, WikiNode) and child.kind == target_kind:
+                yield child
+
+    def _find_node_recursively(
+        self, current_node: Union["WikiNode", str], target_kind: NodeKind
+    ) -> Iterator["WikiNode"]:
+        if isinstance(current_node, WikiNode):
+            for child in current_node.children:
+                if isinstance(child, WikiNode):
+                    if child.kind == target_kind:
+                        yield child
+                    for nest_child in child.children:
+                        yield from self._find_node_recursively(
+                            nest_child, target_kind
+                        )
+
+    def find_child_recursively(
+        self, target_kind: NodeKind
+    ) -> Iterator["WikiNode"]:
+        yield from self._find_node_recursively(self, target_kind)
+
+    def contain_node(self, target_kind: NodeKind) -> bool:
+        for node in self.find_child_recursively(target_kind):
+            return True
+        return False
 
 class TemplateNode(WikiNode):
     def __init__(self, linenum: int):
         super().__init__(NodeKind.TEMPLATE, linenum)
-        self._template_parameters: Optional[Dict[str | int, str | WikiNode]] = None
+        self._template_parameters: Optional[
+            Dict[str | int, str | WikiNode]
```

Comment:
We haven't yet upgraded to a version of Python that support the pipe for type annotations. Please use Union[]

Reply:
I forget to use Union... Actually we can use `|`, it was introduced in Python 3.10.

Reply:
Maybe we could consider using [container](https://podman.io) or [pyenv](https://github.com/pyenv/pyenv) on the server, using Python 3.11 is the easiest way to improve speed: https://docs.python.org/3/whatsnew/3.11.html#faster-cpython

Reply:
I was wrong when I said the dev machine was using Python 3.8, that was my own local machine... The dev is on 3.10.

However, it is unlikely we will upgrade to 3.11 any time soon.